test/decl/protocol/associated_type_witness_availability.swift

```diff
@@ -0,0 +1,83 @@
+// RUN: %swift -typecheck -verify -target %target-cpu-apple-macosx12 %s
+// REQUIRES: OS=macosx
+
+protocol ProtoWithAssocType {
```

Comment:
Can you add a test where the witness is inferred, just in case?

```
protocol ProtoWithAssocType {
  associatedtype A
  func f(_: A)
}


@available(...) struct SomeType {}

struct S: ProtoWithAssocType {
  func f(_: SomeType)
}

Reply:
Good catch, inference will need to diagnose when the availability of the inferred type is insufficient I guess.

Reply:
If these availability checks had been around from the beginning I suppose it's possible that they would have been implemented in associated type inference to reject the (potentially) unavailable candidates and look for others. But I think I'm going to keep these checks after associated type inference for now because it's going to be a lot to move them into inference. I've also discovered that there are source compatibility issues with existing frameworks, so the diagnostics need to be downgraded to warnings in existing language modes, which would be extra messy to do in type inference. To successfully diagnose the inference cases I just need to look through the inferred type to its original declaration, instead of checking the availability of the direct witness decl, which may be a synthesized typealias.